web/application.py

```diff
@@ -310,40 +305,38 @@ def peep(iterator):
             return itertools.chain([firstchunk], iterator)
 
         def wsgi(env, start_resp):
-            # clear threadlocal to avoid inteference of previous requests
-            self._cleanup()
```

Comment:
What are the implications of not cleaning up the threadlocal?

Reply:
The clean up has been moved into the `guard_ctx()` context manager that is used below.

Before this change, the thread local was `.clear()`'d multiple times per request: once at the start of `.wsgi()`, once in `.load()` and once at the end of `.wsgi()` (when the generator is exhausted).  It is now done exactly twice: when entering and leaving the `guard_ctx()` context manager.

Reply:
> Since this looks like a very special case, can this be implemented as an optional middleware that can be enabled explicitly?

Yes, this can be done *externally* to the application (e.g with WSGI middleware that encloses the `web.py` WSGI app callable).  I already do this as a workaround in my application.

However, I wasted a lot of time figuring out `web.py` internals to figure out what was going on and how to circumvent the issue and I find my fix quite hacky.  Issue #268 mentions the same exception (although the conditions that trigger it are different) and this change solves both of the problems.

I though it might save other people some time down the road to fix it in the framework directly...

Reply:
> The clean up has been moved into the `guard_ctx()` context manager that is used below.
> 
> Before this change, the thread local was `.clear()`'d multiple times per request: once at the start of `.wsgi()`, once in `.load()` and once at the end of `.wsgi()` (when the generator is exhausted). It is now done exactly twice: when entering and leaving the `guard_ctx()` context manager.

Before this change, the cleanup was clearing all threadlocal storage objects, but after this change it is clearning only `app.ctx`. There are other places `ThreadedDict` is used. One such place is `web.db` and the users can also use an instance of ThreadedDict in thier code. I've added a sample program in a comment below to demonstrate that the new cleanup mechanism is not enough and not backward-compatible.

Reply:
Sorry, my bad!  This requirement is not stated in the documentation and there are no unit tests to verify this behavior.  I can update the merge request to clear all `web.theadeddict` objects instead (and possibly add a couple of tests to help prevent this form of breakage in the future).

I guess this also means `guard_ctx()` should save all thread-local storage objects like it does for `web.ctx`?